WordPress/openverse
# Add `catalog_cleaner` DAG

## Fixes



Fixes #3415 by @krysal 

## Description


This PR adds a DAG similar to the `bactched_update` but is dedicated to using the files from AWS S3 generated from the ingestion server. I created a new DAG since the `batched_update` would need more tweaks than just loading the data from S3 into a table since the update is performed by joining the newly created table with `image`, and also does the batch updates in parallel using the Dynamic Task Mapping feature of Airflow. Given several `batched_update` runs have been performed with the popularity calculations, it suggests it's possible to parallelize this task as long as the rows to be modified do not overlap. The ingestion server also similarly applies the updates.

This is a proof of concept to see the feasibility of parallelizing the DB updates, making a separate DAG we don't interfere with those that depend on the `batched_update` DAG, and if it's successful (which I don't see why it wouldn't be) we can discuss how to generalize it, whether if modiying the `batched_update` of creating another to use with temporary tables from S3 files.

## Testing Instructions

You will need to have some "dirty" rows in the catalog and their corresponding TSV files containing the correct data to fix them. 

1. To create a TSV, spin up the catalog stack, `./ov just catalog/up`, and if you have `psql` locally try this:
```sh
PGPASSWORD=deploy psql -h localhost -p 50255 openledger deploy -c "COPY (SELECT identifier, url FROM image ORDER BY created_on DESC LIMIT 15) TO STDOUT WITH DELIMITER E'\t' CSV;" >> url.tsv
```
You should then have a file of `<identifier>	<url>` without quotes or a header like the following:
```
3074629a-9934-464b-aca6-b517f4b7cf80	https://pd.w.org/2024/06/8616676f5d15f44f7.79186143-2048x1366.jpg
e3637802-7ff5-4972-b1b4-554e97d921f1	https://pd.w.org/2024/06/85667a9763e32215.10445933-1152x2048.jpg
aec4bdef-6315-4a5d-a997-727891d2f00a	https://pd.w.org/2024/06/121666b021b681b06.28758133-1536x2048.jpg
...
```

2. Then modify the URLs for the same rows. You can use a query like the following to remove the protocol:
```sql
UPDATE image SET url = ltrim(url, 'https://') WHERE identifier IN (
	SELECT identifier FROM image ORDER BY created_on DESC LIMIT 15
);
```
3. Upload your new file to MinIO: http://localhost:5011/. Save the path to the bucket as you'll need it as a setting later.

4. Go to the Airflow U, unpause the `catalog_cleaner` DAG, and trigger it with a configuration that fits your test case. E.g.:
```json
{
    "batch_size": 3,
    "column": "url",
    "s3_bucket": "openverse-catalog",
    "s3_path": "shared/data-refresh-cleaned-data/url_test.tsv"
}
```

5. Verify the changes were applied in the DB to the rows.
```sh
./ov just catalog/pgcli
```

```sql
SELECT identifier, url FROM image ORDER BY created_on DESC LIMIT 15;
```

Try different configurations for `batch_size` and/or the `CLEANER_MAX_CONCURRENT_DB_UPDATE_TASKS` Airflow variable to see how it works. You can add `time.sleep(seconds)`  before returning in the `update_batch` function to better appreciate the concurrency.

## Checklist



[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
